File: runme/mcgen/local/datan/solsrc/s2ml.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Table of constant values */

static integer c__75 = 75;
static double c_d1 = 1.;

main(void)
{
    /* Local variables */
    static integer i, n, iexp, lcapt, nexp, npt, nws, nx;
    static double hist[100], r, x[2], sigma[2], dplus[4] /* was [2][2] */,
                  x0, rho, sprime[2], spsq[2], a[2], c[4] /* was [2][2] */,
                  xbar[2], delx /*, xsampl[20000]*/ /* was [10000][2] */;
#if defined(_MSC_VER)
    static double __huge xsampl[20000];
#elif defined(__TURBOC__)
    static double __huge *xsampl;
#else
    static double xsampl[20000];
#endif
    static char capt[75], tx[75], ty[75];
    /* System generated locals */
    integer i__1, i__2;
    double d__1, d__2;
/* identify program to user */
    printf("%s\n", "Program S2MLC program plots histogram");
    printf("%s\n\n", "of sample correlation coefficient.");
/* ask for input */
    printf("%s\n", "Enter number NEXP of simulated experiments (>>1):");
    printf("%s", "> ");
    scanf("%li", &nexp);
    printf("%s\n", "Enter number NPT of points in each experiment (>2):");
    printf("%s", "> ");
    scanf("%li", &npt);
#if defined(_MSC_VER)
/*
    xsampl = (double __huge *)_halloc(2L * npt, sizeof(double));
*/
#elif defined(__TURBOC__)
    xsampl = (double __huge *)farmalloc(2L * npt * sizeof(double));
#endif
    if (!xsampl) {
       MEMERROR("s2ml: xsampl");
       return(0);
    }
    printf("%s\n", "Enter a1:");
    printf("%s", "> ");
    scanf("%lf", &a[0]);
    printf("%s\n", "Enter a2:");
    printf("%s", "> ");
    scanf("%lf", &a[1]);
    printf("%s\n", "Enter sigma1 (>0.):");
    printf("%s", "> ");
    scanf("%lf", &sigma[0]);
    printf("%s\n", "Enter sigma2 (>0.):");
    printf("%s", "> ");
    scanf("%lf", &sigma[1]);
    printf("%s\n", "Enter rho (-1.<rho<1.):");
    printf("%s", "> ");
    scanf("%lf", &rho);
/* N=2 defines number of variables in multivariate normal distribution */
    n = 2;
/* set up covariance matrix */
/* Computing 2nd power */
    d__1 = sigma[0];
    c[0] = d__1 * d__1;
/* Computing 2nd power */
    d__1 = sigma[1];
    c[3] = d__1 * d__1;
    c[2] = rho * sigma[0] * sigma[1];
    c[1] = c[2];
/* prepare for generation of random numbers */
    rnmnpr_(c, dplus, &n);
/* initialize histogram */
    x0 = -1.5;
    delx = .03f;
    nx = 100;
    smhsin_(hist, &x0, &delx, &nx);
/* loop over all simulation experiments */
    i__1 = nexp;
    for (iexp = 1; iexp <= i__1; ++iexp) {
/* generate sample of 2-vectors from bivariate normal */
/* and compute sample means */
        xbar[0] = 0.;
        xbar[1] = 0.;
        i__2 = npt;
        for (i = 1; i <= i__2; ++i) {
            rnmngn_(dplus, a, x, &n);
            xsampl[i - 1] = x[0];
            xsampl[i + npt - 1] = x[1];
            xbar[0] += x[0];
            xbar[1] += x[1];
        }
        xbar[0] /= (double) npt;
        xbar[1] /= (double) npt;
/* compute sample variances and correlation coefficient */
        spsq[0] = 0.;
        spsq[1] = 0.;
        r = 0.;
        i__2 = npt;
        for (i = 1; i <= i__2; ++i) {
/* Computing 2nd power */
            d__1 = xsampl[i - 1] - xbar[0];
            spsq[0] += d__1 * d__1;
/* Computing 2nd power */
            d__2 = xsampl[i + npt - 1] - xbar[1];
            spsq[1] += d__2 * d__2;
            r += d__1 * d__2;
        }
        spsq[0] /= (double) npt;
        spsq[1] /= (double) npt;
        sprime[0] = sqrt(spsq[0]);
        sprime[1] = sqrt(spsq[1]);
        r /= (double) npt * sprime[0] * sprime[1];
        smhsfl_(hist, &x0, &delx, &nx, &r, &c_d1);
    }
/* prepare texts */
    memset(tx, ' ', 75);
    memset(ty, ' ', 75);
    strncpy(tx, "r", 1);
    strncpy(ty, "N(r)", 4);
    sprintf(capt,
            "a_1#=%6.2lf, a_2#=%6.2lf, &s_1#=%5.2lf, &s_2#=%5.2lf, &r=%5.2lf",
            a[0], a[1], sigma[0], sigma[1], rho);
    lcapt=strlen(capt);
/* graphical output */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    (void)getchar();
    smhsgr_(hist, &x0, &delx, &nx, tx, &c__75, ty, &c__75, capt, &lcapt, &nws);
#if defined(_MSC_VER)
/*
    _hfree((void __huge *) xsampl);
*/
#elif defined(__TURBOC__)
    farfree((void far *) xsampl);
#endif
    return 0;
} /* main */

File: runme/mcgen/local/datan/solsrc/s2ml.f
      PROGRAM S2ML
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      DIMENSION C(2,2),DPLUS(2,2),X(2),A(2),SIGMA(2)
      DIMENSION HIST(100)
      CHARACTER*75 TX,TY,CAPT
      DIMENSION XSAMPL(10000,2),XBAR(2),SPSQ(2),SPRIME(2)
      PARAMETER(ZERO=0.D0)
C identify program to user
      WRITE(*,*)' Program S2MLC program plots histogram'
      WRITE(*,*)' of sample correlation coefficient'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)' Enter number NEXP of simulated experiments (>>1)'
      WRITE(*,*)' >'
      READ(*,*)NEXP
      WRITE(*,*)
     +' Enter number NPT of points in each experiment (3<=NPT<=10000)'
      WRITE(*,*)' >'
      READ(*,*)NPT
      WRITE(*,*)' Enter a1'
      WRITE(*,*)' >'
      READ(*,*)A(1)
      WRITE(*,*)' Enter a2'
      WRITE(*,*)' >'
      READ(*,*)A(2)
      WRITE(*,*)' Enter sigma1 (>0.)'
      WRITE(*,*)' >'
      READ(*,*)SIGMA(1)
      WRITE(*,*)' Enter sigma2 (>0.)'
      WRITE(*,*)' >'
      READ(*,*)SIGMA(2)
      WRITE(*,*)' Enter rho (-1.<rho<1.)'
      WRITE(*,*)' >'
      READ(*,*)RHO
C N=2 defines number of variables in multivariate normal distribution
      N=2
C set up covariance matrix
      C(1,1)=SIGMA(1)**2
      C(2,2)=SIGMA(2)**2
      C(1,2)=RHO*SIGMA(1)*SIGMA(2)
      C(2,1)=C(1,2)
C prepare for generation of random numbers
      CALL RNMNPR(C,DPLUS,N)
C initialize histogram
      X0=-1.5D0
      DELX=0.03
      NX=100
      CALL SMHSIN(HIST,X0,DELX,NX)
C loop over all simulation experiments
      DO 30 IEXP=1,NEXP
C generate sample of 2-vectors from bivariate normal
C and compute sample means
        XBAR(1)=ZERO
        XBAR(2)=ZERO
        DO 10 I=1,NPT
          CALL RNMNGN(DPLUS,A,X,N)
          XSAMPL(I,1)=X(1)
          XSAMPL(I,2)=X(2)
          XBAR(1)=XBAR(1)+X(1)
          XBAR(2)=XBAR(2)+X(2)
10      CONTINUE
        XBAR(1)=XBAR(1)/DBLE(NPT)
        XBAR(2)=XBAR(2)/DBLE(NPT)
C compute sample variances and correlation coefficient
        SPSQ(1)=ZERO
        SPSQ(2)=ZERO
        R=ZERO
        DO 20 I=1,NPT
          SPSQ(1)=SPSQ(1)+(XSAMPL(I,1)-XBAR(1))**2
          SPSQ(2)=SPSQ(2)+(XSAMPL(I,2)-XBAR(2))**2
          R=R+(XSAMPL(I,1)-XBAR(1))*(XSAMPL(I,2)-XBAR(2))
20      CONTINUE
        SPSQ(1)=SPSQ(1)/DBLE(NPT)
        SPSQ(2)=SPSQ(2)/DBLE(NPT)
        SPRIME(1)=SQRT(SPSQ(1))
        SPRIME(2)=SQRT(SPSQ(2))
        R=R/(DBLE(NPT)*SPRIME(1)*SPRIME(2))
        CALL SMHSFL(HIST,X0,DELX,NX,R,1.D0)
30    CONTINUE
C prepare texts
      TX='r'
      TY='N(r)'
      WRITE(CAPT(1:60),'(A)')
     +'a_1#=******, a_2#=******, &s_1#=*****, &s_2#=*****, &r=*****'
      WRITE(CAPT(6:11),'(F6.2)')A(1)
      WRITE(CAPT(19:24),'(F6.2)')A(2)
      WRITE(CAPT(33:37),'(F5.2)')SIGMA(1)
      WRITE(CAPT(46:50),'(F5.2)')SIGMA(2)
      WRITE(CAPT(56:60),'(F5.2)')RHO
C graphical output
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
      CALL SMHSGR(HIST,X0,DELX,NX,TX,75,TY,75,CAPT,60,NWS)
      END
